scripts/state-surgery.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * State Surgery Script
+ *
+ * Required env vars:
+ *  SEQUENCER_ENDPOINT
+ *  STATE_DUMP_PATH
+ */
+
+const fs = require('fs');
+const axios = require('axios');
+const { JsonRpcProvider } = require('@ethersproject/providers');
+const { getContractDefinition } = require('@eth-optimism/contracts')
+
+const cfg = config()
+
+const contracts = {
+  ProxyEOA: getContractDefinition('OVM_ProxyEOA'),
+  Lib_AddressManager: getContractDefinition('Lib_AddressManager'),
+  OVM_DeployerWhitelist: getContractDefinition('OVM_DeployerWhitelist'),
+  OVM_L1MessageSender: getContractDefinition('OVM_L1MessageSender'),
+  OVM_L2ToL1MessagePasser: getContractDefinition('OVM_L2ToL1MessagePasser'),
+  OVM_ProxyEOA: getContractDefinition('OVM_ProxyEOA'),
+  OVM_ECDSAContractAccount: getContractDefinition('OVM_ECDSAContractAccount'),
+  mockOVM_ECDSAContractAccount: getContractDefinition('mockOVM_ECDSAContractAccount'),
+  OVM_ProxySequencerEntrypoint: getContractDefinition('OVM_ProxySequencerEntrypoint'),
+  ERC1820Registry: getContractDefinition('ERC1820Registry'),
+  OVM_SequencerEntrypoint: getContractDefinition('OVM_SequencerEntrypoint'),
+  OVM_L2CrossDomainMessenger: getContractDefinition('OVM_L2CrossDomainMessenger'),
+  OVM_SafetyChecker: getContractDefinition('OVM_SafetyChecker'),
+  OVM_ExecutionManager: getContractDefinition('OVM_ExecutionManager'),
+  OVM_StateManager: getContractDefinition('OVM_StateManager'),
+  OVM_ETH: getContractDefinition('OVM_ETH'),
+}
+
+const sequencer = new JsonRpcProvider(cfg.sequencerEndpoint);
+
+;(async () => {
+  let currentState;
+
+  while (!currentState) {
+    try {
+      currentState = await sequencer.send('debug_dumpBlock', ['latest']);
+    } catch (e) {
+      // Do nothing
+    }
+  }
+
+  // Need to merge current state into contractsDump
+  const res = await axios.get(cfg.stateDumpPath)
+  const contractsDump = res.data
+
+  for (const [address, account] of Object.entries(currentState.accounts)) {
+    // EOA Accounts receive the latest OVM_ProxyEOA code. They keep the same
+    // storage and nonce.
+    if (isEOA(account)) {
+      const eoaName = 'EOA_' + address
+      contractsDump.accounts[eoaName] = {
+        address: address,
+        nonce: account.nonce,
+        code: contracts.ProxyEOA.deployedBytecode,
+        storage: account.storage,
+        abi: contracts.ProxyEOA.abi
+      }
+    } else if (isPredeploy(address) || isSystemAccount(address)) {
+      // Predeploys and System Accounts keep the same nonce and code
+      // Iterate through the contractsDump to find the matching addresses so
+      // that the storage can be pulled in
+      for (const [name, dumpAccount] of Object.entries(contractsDump.accounts)) {
```

Comment:
are we worried about the performance of this double loop, or is there not a lot of these?

Reply:
This ended up not being required